samples/third-party/gltf-loader.js

```diff
@@ -0,0 +1,354 @@
+(function() {
+    'use strict';
+
+    var GLTF = {};
+
+    var AttributeSize = {
+        5120: 1, // BYTE
+        5121: 1, // UNSIGNED_BYTE
+        5122: 2, // SHORT
+        5123: 2, // UNSIGNED_SHORT
+        5126: 4  // FLOAT
+    };
+
+    var NumberOfComponents = {
+        'SCALAR': 1,
+        'VEC2': 2,
+        'VEC3': 3,
+        'VEC4': 4,
+        'MAT2': 4,
+        'MAT3': 9,
+        'MAT4': 16
+    };
+
+    // @todo: load multiple scences
+    GLTF.Scene = function() {
+        this.vertices = [];
+        this.indices = [];
+    };
+
+    var parseGltf = function(json, onload) {
+
+        var newScene = new GLTF.Scene;
+
+        // Iterate through every scene
+        for (var sceneName in json.scenes) {
+            var scene = json.scenes[sceneName];
+            var nodes = scene.nodes;
+            var nodeLen = nodes.length;
+
+            // Iterate through every node within scene
+            for (var n = 0; n < nodeLen; ++n) {
+                var nodeName = nodes[n];
+                var node = json.nodes[nodeName];
+
+                // Traverse node
+                traverseNode(json, node, newScene, onload);
+            }
+        }
+    };
+    
+    
+    var translationVec3;
+    var rotationQuat;
+    var scaleVec3;
+    var TRMatrix;
+    
+    var traverseNode = function(json, node, scene, onload, matrix = mat4.create()) {
+
+        var curMatrix = mat4.create();
+        
+        if (node.hasOwnProperty('matrix')) {
+            // matrix
+            for(var i = 0; i < 16; ++i) {
+                curMatrix[i] = node.matrix[i];
+            }
+            mat4.multiply(curMatrix, matrix, curMatrix);
+        } else {
+            // translation, rotation, scale (TRS)
+            // @todo: these labels are optional
+            vec3.set(translationVec3, node.translation[0], node.translation[1], node.translation[2]);
+            quat.set(rotationQuat, node.rotation[0], node.rotation[1], node.rotation[2], node.rotation[3]);
+            mat4.fromRotationTranslation(TRMatrix, rotationQuat, translationVec3);
+            mat4.multiply(curMatrix, curMatrix, TRMatrix);
+            vec3.set(scaleVec3, node.scale[0], node.scale[1], node.scale[2]);
+            mat4.scale(curMatrix, curMatrix, scaleVec3);
+        }
+            
+        // Iterate through every mesh within node
+        var meshes = node.meshes;
+        if(!!meshes) {
+            var meshLen = meshes.length;
+            for (var m = 0; m < meshLen; ++m) {
+                var meshName = meshes[m];
+                var mesh = json.meshes[meshName];
+
+                // Iterate through primitives
+                var primitives = mesh.primitives;
+                var primitiveLen = primitives.length;
+
+                for (var p = 0; p < primitiveLen; ++p) {
+                    var primitive = primitives[p];
+
+                    // Get indices first
+                    parseIndices(json, primitive, scene, function() {
```

Comment:
Since we are doing a top down loading approach (scene -> nodes -> meshes -> primitives -> ... buffers) as opposed to a bottom up loading approach where we would load the buffers first, make sure to use caching so we don't load the same buffer twice or duplicate any data.